programs/gumball-machine/src/lib.rs

```diff
@@ -96,10 +150,103 @@ pub struct Destroy<'info> {
     authority: Signer<'info>,
 }
 
+#[inline(always)]
+// For efficiency, this returns the GumballMachineHeader because it's required to validate
+// payment parameters. But the main purpose of this function is to determine which config
+// line to mint to the user, and CPI to bubblegum to actually execute the mint
+fn find_and_mint_compressed_nft<'info>(
+    gumball_machine: &AccountInfo<'info>,
+    payer: &Signer<'info>,
+    willy_wonka: &AccountInfo<'info>,
+    willy_wonka_bump: &u8,
+    recent_blockhashes: &UncheckedAccount<'info>,
+    instruction_sysvar_account: &UncheckedAccount<'info>,
+    bubblegum_authority: &AccountInfo<'info>,
+    nonce: &AccountInfo<'info>,
+    gummyroll: &Program<'info, Gummyroll>,
+    merkle_slab: &AccountInfo<'info>,
+    bubblegum: &Program<'info, Bubblegum>,
+    num_items: u64
+) -> Result<GumballMachineHeader> {
+    // Load all data
+    let mut gumball_machine_data = gumball_machine.try_borrow_mut_data()?;
+    let (mut header_bytes, config_data) =
+        gumball_machine_data.split_at_mut(std::mem::size_of::<GumballMachineHeader>());
+    let gumball_header = GumballMachineHeader::load_mut_bytes(&mut header_bytes)?;
+    let clock = Clock::get()?;
+    assert!(clock.unix_timestamp > gumball_header.go_live_date);
+    let size = gumball_header.max_items as usize;
+    let index_array_size = std::mem::size_of::<u32>() * size;
+    let config_size = gumball_header.extension_len * size;
+    let line_size = gumball_header.extension_len;
+
+    assert!(config_data.len() == index_array_size + config_size);
+    let (indices_data, config_lines_data) = config_data.split_at_mut(index_array_size);
+
+    // TODO: Validate data
```

Comment:
this step should be added as a function + test

Reply:
Not sure what this was referring too tbh, just refactored this instr body into helper. @jarry-xiao what were we looking to validate here?